Automated cherry pick of #109694: Be sure to update the status of StatefulSet even if the new

gjkim42 opened the pull request:
ref: https://github.com/kubernetes/kubernetes/issues/112010

Cherry pick of #109694 #112737 on release-1.23.

#109694: Be sure to update the status of StatefulSet even if the new
#112737: StatefulSet: Cleanup the complex defer function updating the

For details on the cherry pick process, see the [cherry pick requests](https://git.k8s.io/community/contributors/devel/sig-release/cherry-picks.md) page.

```release-note
Fixed StatefulSet to show the valid status even if the new replica creation fails.
```

RyanAoh committed:
Be sure to update the status of StatefulSet even if the new replica creation fails

gjkim42 commented:
/retest

soltysh commented:
/hold
see https://github.com/kubernetes/kubernetes/pull/109694/files#r979974744

gjkim42 committed:
StatefulSet: Cleanup the complex defer function updating the status

In the long term, the complex defer function makes the code harder to
maintain as code after it should take that into account. This removes
the complex defer function updating the status of a statefulset.

gjkim42 commented:
@soltysh @atiratree 
cherry-picked #112737 to resolve https://github.com/kubernetes/kubernetes/pull/112084#issuecomment-1257966715.

Could you PTAL?

gjkim42 committed:
Fix a conflict

aojea commented:
/hold

the tests added with this PR are flaking in master, we should verify this is not going to regress

https://github.com/kubernetes/kubernetes/issues/113036

atiratree commented:
The flaking tests were added by a different PR (https://github.com/kubernetes/kubernetes/pull/112127) for daemon sets. This PR should be ok,  but yes we should wait with backporting for some time to see that this does not cause any regressions

atiratree commented:
retesting flakes
/retest

atiratree commented:
we can proceed here, as the code seems to be stable
/hold cancel

soltysh reviewed (approved):
/lgtm
/approve

soltysh commented:
/triage accepted
/priority important-longterm
/kind bug

gjkim42 commented:
/cc @kubernetes/release-managers

Verolop commented:
/lgtm

Verolop commented:
/test pull-kubernetes-integration
